### PR TITLE
check for invalid hook function signatures

### PR DESCRIPTION
**Pull request by brettmc**
trying to execute a hook function when the provided callback's signature does not match reality is a fatal error. Check the signature before calling, and if they do not match, do not execute the callback

Fixes: https://github.com/open-telemetry/opentelemetry-php-instrumentation/issues/81

### CODE REVIEW

**Comment by brettmc** [comment]
This is a different approach to #56 - instead of checking at hook/init time, it checks is runtime just prior to actually calling the pre/post hook. The benefit of doing it here is that we can do extra type checking on the class (pre+post), and exception (post).
I've chosen to make it silently not execute an incompatible callback (since it's runtime), but we could do this in conjunction with #56 which I think could complain more loudly (throw an exception?).

**Comment by pdelewski** [comment]
> This is a different approach to #56 - instead of checking at hook/init time, it checks is runtime just prior to actually calling the pre/post hook. The benefit of doing it here is that we can do extra type checking on the class (pre+post), and exception (post). I've chosen to make it silently not execute an incompatible callback (since it's runtime), but we could do this in conjunction with #56 which I think could complain more loudly (throw an exception?).

@brettmc Just to clarify, so this PR and #56 are complementary and #56 still make sense from your perspective, correct?

**Comment by pdelewski** [comment]
@brettmc Overall, LGTM, however spent only few minutes looking at the code (still vacationing).

**Comment by brettmc** [comment]
> Just to clarify, so this PR and https://github.com/open-telemetry/opentelemetry-php-instrumentation/pull/56 are complementary and https://github.com/open-telemetry/opentelemetry-php-instrumentation/pull/56 still make sense from your perspective, correct?

I think that they can be complementary. #56 is a startup check, but it can't catch all issues. I think that this PR will stop all fatals from invalid signature, but because it is a runtime check, it has to do it quietly.

File: ext/otel_observer.c [hunk 1]
```diff
@@ -163,6 +213,10 @@ static void observer_begin(zend_execute_data *execute_data, zend_llist *hooks) {
         fci.named_params = NULL;
         fci.retval = &ret;
 
+        if (!is_valid_signature(fci, fcc)) {
+            continue;
```
[the remark this answers is not in the document]
Comment (brettmc) [reply ext/otel_observer.c:238]: now it logs a message to error_log. And, can be disable via .ini setting

File: ext/otel_observer.c [hunk 2]
```diff
@@ -133,6 +133,56 @@ static inline void func_get_lineno(zval *zv, zend_execute_data *ex) {
...
+        // get type mask of callback argument
+        arg_info = &func->common.arg_info[i];
+        arg_type = &arg_info->type;
+        type_mask = arg_type->type_mask;
+
+        // get actual value + type
+        zval param = fci.params[i];
+        type = Z_TYPE(fci.params[i]);
+
+        if (arg_type->type_mask == 0) {
```
[the remark this answers is not in the document]
Comment (brettmc) [reply ext/otel_observer.c]: I think it is, but I've used a const instead. Basically, if no type hint, there is nothing to check.

File: ext/opentelemetry.c [hunk 3]
```diff
@@ -71,11 +69,22 @@ static int check_conflicts() {
         conflict_found = 1;
     }
 
-    return conflict_found;
+    OTEL_G(disabled) = conflict_found;
```
[the remark this answers is not in the document]
Comment (brettmc) [reply ext/opentelemetry.c:72]: yes you are correct. updated!

**Review by pdelewski — APPROVED** [review]

**Pull request merged**